### DISCUSSION TITLE
Optimise workflows?

### DISCUSSION THREAD

**Discussion by tothxa** [widelands/widelands#5880] [General]
@widelands/developers

It looks like we can run fewer workflow jobs in parallel than before, so I wonder if we could make better use of the available resources and speed up PR testing by optimising our workflows.

My ideas for optimisation so far:

 - Cache build directories and only do partial rebuilds for PR builds (I guess master should always be rebuilt cleanly) (I'm not sure how much cache space this would use, and how much we have)
 - Better yet, when nothing changes in `src/` then don't even rebuild, and only re-run regression tests if `data/` or `test/` changed
 - Try to run `clang-tidy` only on changed files? Though headers are difficult, because we need to track down all places where they are included, even transitively… (again, master would still get full check)

Does this make sense? Do you have more ideas?

**Comment by matthiakl** [comment]
> Cache build directories and only do partial rebuilds for PR builds (I guess master should always be rebuilt cleanly) (I'm not sure how much cache space this would use, and how much we have)

The cache limit is 10GB ([source](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy)). A full debug build creates a directory of about 3GB, however, half of it are the linked executables. So it would help to run `find build -type f -executable -delete` before creating the cache and do the link step every time.
The cache is compressed with zst which reduces the size to ~450MB. So it should be possible to cache at least the build directories of the linux builds.
The workflow could be:
1. build master from scratch
2. run tests and upload artifacts
3. delete the executables
4. store build directory as cache
5. In a PR: use cache from master, but do not save it after build

> Better yet, when nothing changes in src/ then don't even rebuild, and only re-run regression tests if data/ or test/ changed

This would mean to cache the executables. I don't think it is worth the cache space.

I am currently working on improving the cache usage regarding the vcpkg builds to reduce the number of required package rebuilds.

**Reply by tothxa** [reply]
I'm now experimenting with ccache locally. I wonder if we could merge the `ccache`s from master and the previous builds of a PR. That would certainly beat anything that make/ninja can ever do as far as compilation is concerned, and we wouldn't have to mess with `mtime`s. OTOH make/ninja can also skip some linking, but I think only the final linking of the main executable is slow. (quick test: after removing all `.a` files, `ninja libwidelands_ball_of_mud.a` took 6.2 seconds real time – yes, I even forgot the `-j3`)

**Reply by tothxa** [reply]
> > Better yet, when nothing changes in src/ then don't even rebuild, and only re-run regression tests if data/ or test/ changed
> 
> This would mean to cache the executables. I don't think it is worth the cache space.

OK, then how about using the last daily build for these? … But we only have appimages. I'm not familiar with them.

**Reply by matthiakl** [reply]
> I wonder if we could merge the ccaches from master and the previous builds of a PR. 

Can you give estimations on the ccache size? 
With some scripting and digging into the ccache structure I think this is possible, if the ccache is considerably smaller then caching the build directory. It certainly be more sensible.

> but I think only the final linking of the main executable is slow.

This is especially true for the static Windows builds. Does ccache can also be used here?

> OK, then how about using the last daily build for these?

Currently we only have clang-10 Release and Debug builds as AppImages. Also they are designed to be standalone, so you can just download and execute it. All data files are packed into the image, but `--datadir=**` should be usable to point to the PR.

**Reply by tothxa** [reply]
> Can you give estimations on the ccache size?

A release build on my debian oldstable system created a 212MB uncompressed cache (total of `*.o` files in the build dir is 207MB), which compressed down with `tar -J` (LZMA) to ~140MB. A debug build created 1.59GB uncompressed cache (total `*.o` is 1.62GB and ccache reports 35 hits? I don't quite understand this part) which compressed down to 191MB.

> > but I think only the final linking of the main executable is slow.
> This is especially true for the static Windows builds. Does ccache can also be used here?

No, it only caches the object files. It doesn't handle linking at all.

> > OK, then how about using the last daily build [when only data files are changed]?
> Currently we only have clang-10 Release and Debug builds

Well, I guess one release and one debug build would be enough for this case. It's rather unlikely that such a change would hit a compiler incompatibility issue, and in the worst case, it would be found by the next full run.

> as AppImages. Also they are designed to be standalone, so you can just download and execute it. All data files are packed into the image, but `--datadir=**` should be usable to point to the PR.

I'm afraid that complicates things for `regression_test.py` to a point where it's not worth the trouble.

**Reply by tothxa** [reply]
I use ccache for all builds locally now, and it saves some time when switching branches in the same worktree, but in the end it does much less than I expected. When I first read about its existence, I thought it was a feature of the compilers themselves, and worked at lower levels. But it turns out all it does is checks whether anything changed in each compile unit (ie. a source file and all its includes), and caches the resulting object files (mostly the same as what I did for clang-tidy in #5922, only they can usually skip running the preprocessor to detect hits). This means that if there's even just a typo change in some low level header file, then a large part of the codebase needs recompiling.

I also checked the testsuite runs again, and build times are actually only 15-20 minutes. It's the regression test suite that takes more than an hour. So with the large cache size it's probably not worth it to cache every build. But maybe it could still be used for 1 or 2 "main" configs that are always built first, and the rest of the matrix would only start if the compilation phase of these succeeds. (that'll need some advanced workflow trickery though). Also these would be used for the regression tests when there are no changes in `src`.

**Comment by tothxa** [comment]
And for clang-tidy, we can try https://github.com/freedick/cltcache

**Reply by tothxa** [reply]
Well, it didn't work for us (it expects compiler arguments on the clang-tidy command line, not in `compile_commands.json`), but I integrated something similar into `run-clang-tidy.py` in #5922.

**Comment by SimonHeimberg** [comment]
Most time is spent in running tests. I did an analysation of duration of tests run for one run, for https://github.com/widelands/widelands/actions/runs/5556166018/jobs/10148384149?pr=6017

shortly my summary:
- (time until workflow was completed: [2:4:56](https://github.com/widelands/widelands/actions/runs/5556166018?pr=6017))
- (time for compilation was 0:13:53)
- total duration running the test suite is 1:24:20 (as reported in the summary)
- the map **all_tribes** has the top 5 of longest tests
- the sum of the duration of map **all_tribes** is 0:43:56, which is more than half of the total time to run the tests
- the output of the top 5 tests tells the duration for two parts of the test
  loading the saved game takes ~52s
  the rest (5min to 8min) is spent before

Here is the duration for each test (total duration is 1:24:20) (table is from calc, I can attach the calc file if desired)

duration | of test | tRel |  rank of duration
-- | -- | -- | --
00:09:25 | test/maps/all_tribes.wmf/scripting/test_amazons.lua | 11.17% | 1
00:09:11 | test/maps/all_tribes.wmf/scripting/test_atlanteans.lua | 10.89% | 2
00:06:28 | test/maps/all_tribes.wmf/scripting/test_barbarians.lua | 7.67% | 5
00:09:09 | test/maps/all_tribes.wmf/scripting/test_empire.lua | 10.85% | 3
00:08:53 | test/maps/all_tribes.wmf/scripting/test_frisians.lua | 10.53% | 4
00:00:50 | test/maps/all_tribes.wmf/scripting/test_place_foreign_buildings.lua | 0.99% | 19
00:02:27 | test/maps/expedition.wmf/scripting/test_cancel_started_expedition_on_ship_one_ship.lua | 2.91% | 7
00:02:22 | test/maps/expedition.wmf/scripting/test_cancel_started_expedition_underway_one_ship.lua | 2.81% | 9
00:02:25 | test/maps/expedition.wmf/scripting/test_cancel_started_expedition_underway_two_ships.lua | 2.87% | 8
00:02:28 | test/maps/expedition.wmf/scripting/test_cancel_when_port_space_was_reached_one_ship.lua | 2.92% | 6
00:02:20 | test/maps/expedition.wmf/scripting/test_cancel_when_port_space_was_reached_two_ships.lua | 2.77% | 10
00:01:37 | test/maps/expedition.wmf/scripting/test_check_transportation_works_one_ship.lua | 1.92% | 13
00:01:36 | test/maps/expedition.wmf/scripting/test_check_transportation_works_two_ships.lua | 1.90% | 14
00:01:13 | test/maps/expedition.wmf/scripting/test_no_double_colonizing.lua | 1.44% | 15
00:01:52 | test/maps/expedition.wmf/scripting/test_ship_movement_controls.lua | 2.21% | 11
00:00:15 | test/maps/expedition.wmf/scripting/test_starting_and_immediately_canceling.lua | 0.30% | 50
00:00:34 | test/maps/expedition.wmf/scripting/test_starting_wait_a_while_cancel.lua | 0.67% | 31
00:00:31 | test/maps/lua_persistence.wmf/scripting/test_persistence.lua | 0.61% | 34
00:00:23 | test/maps/lua_testsuite.wmf/scripting/test_lua_in_game.lua | 0.45% | 38
00:00:31 | test/maps/market_trading.wmf/scripting/test_custom_units.lua | 0.61% | 34
00:00:52 | test/maps/market_trading.wmf/scripting/test_foreign_units.lua | 1.03% | 18
00:00:30 | test/maps/market_trading.wmf/scripting/test_market_can_be_built.lua | 0.59% | 36
00:00:20 | test/maps/market_trading.wmf/scripting/test_simple_trade.lua | 0.40% | 42
00:00:14 | test/maps/place_buildings.wmf/scripting/test_place_buildings.lua | 0.28% | 52
00:00:17 | test/maps/plain.wmf/scripting/test_animations.lua | 0.34% | 46
00:00:13 | test/maps/plain.wmf/scripting/test_campaign_data.lua | 0.26% | 58
00:00:18 | test/maps/plain.wmf/scripting/test_casern.lua | 0.36% | 44
00:00:16 | test/maps/plain.wmf/scripting/test_fires_do_not_burn_forever.lua | 0.32% | 48
00:00:13 | test/maps/plain.wmf/scripting/test_gamespeed.lua | 0.26% | 58
00:00:21 | test/maps/plain.wmf/scripting/test_geologist_messages_do_not_crash.lua | 0.42% | 41
00:00:49 | test/maps/plain.wmf/scripting/test_inputqueues.lua | 0.97% | 20
00:00:17 | test/maps/plain.wmf/scripting/test_legacy_animation_dir.lua | 0.34% | 46
00:00:14 | test/maps/plain.wmf/scripting/test_player_vision.lua | 0.28% | 52
00:00:18 | test/maps/plain.wmf/scripting/test_produced_wares_count.lua | 0.36% | 44
00:00:13 | test/maps/plain.wmf/scripting/test_text_renderer_does_not_crash.lua | 0.26% | 58
00:00:44 | test/maps/plain.wmf/scripting/test_tribes_postloading_works.lua | 0.87% | 21
00:00:35 | test/maps/plain.wmf/scripting/test_ui.lua | 0.69% | 30
00:01:38 | test/maps/plain.wmf/scripting/test_ui_2.lua | 1.94% | 12
00:00:14 | test/maps/plain.wmf/scripting/test_upgraded_workers.lua | 0.28% | 52
00:00:13 | test/maps/port_space.wmf/scripting/test_attack_forbidden.lua | 0.26% | 58
00:00:23 | test/maps/port_space.wmf/scripting/test_detection_with_allowed_player_immos.lua | 0.45% | 38
00:00:23 | test/maps/port_space.wmf/scripting/test_portfield_barely_unowned.lua | 0.45% | 38
00:00:14 | test/maps/port_space.wmf/scripting/test_portfield_blocked_by_rock.lua | 0.28% | 52
00:00:40 | test/maps/port_space.wmf/scripting/test_portfield_covered_by_tree.lua | 0.79% | 22
00:00:14 | test/maps/port_space.wmf/scripting/test_portfield_neighbour_owned_by_p2.lua | 0.28% | 52
00:00:14 | test/maps/port_space.wmf/scripting/test_portfield_owned_by_p2.lua | 0.28% | 52
00:00:15 | test/maps/ship_transportation.wmf/scripting/test_create_request_while_worker_with_ware_is_in_transit.lua | 0.30% | 50
00:00:37 | test/maps/ship_transportation.wmf/scripting/test_many_ships.lua | 0.73% | 26
00:00:16 | test/maps/ship_transportation.wmf/scripting/test_rip_farm_with_ware_and_worker_in_transit.lua | 0.32% | 48
00:00:37 | test/maps/ship_transportation.wmf/scripting/test_rip_first_port_with_ware_in_portdock.lua | 0.73% | 26
00:00:37 | test/maps/ship_transportation.wmf/scripting/test_rip_first_port_with_worker_in_portdock.lua | 0.73% | 26
00:00:36 | test/maps/ship_transportation.wmf/scripting/test_rip_portdock_with_worker_and_ware_in_transit.lua | 0.71% | 29
00:00:38 | test/maps/ship_transportation.wmf/scripting/test_rip_ports_with_ware_in_transit.lua | 0.75% | 24
00:00:38 | test/maps/ship_transportation.wmf/scripting/test_rip_ports_with_worker_in_transit.lua | 0.75% | 24
00:00:40 | test/maps/ship_transportation.wmf/scripting/test_rip_second_port_with_ware_in_portdock.lua | 0.79% | 22
00:00:34 | test/maps/ship_transportation.wmf/scripting/test_rip_second_port_with_worker_in_portdock.lua | 0.67% | 31
00:01:04 | test/maps/ship_transportation.wmf/scripting/test_rip_ship_before_picking_up_transporting_ware.lua | 1.26% | 16
00:01:04 | test/maps/ship_transportation.wmf/scripting/test_rip_ship_while_transporting_ware.lua | 1.26% | 16
00:00:32 | test/maps/two_ponds.wmf/scripting/test_seafaring.lua | 0.63% | 33
00:00:19 | test/maps/Aquila.wmf/scripting/editor_test_loading.lua | 0.38% | 43
00:00:26 | test/maps/lua_testsuite.wmf/scripting/editor_test_lua_in_editor.lua | 0.51% | 37

**Comment by SimonHeimberg** [comment]
From #6020

**Current behaviour**
Tests are also run if only a file changed which has nothing to do with the code
example: https://github.com/widelands/widelands/pull/6019

**Expected behaviour**
I suggest to skip tests on obvious cases by using ...

either `paths` or `paths-ignore` for the entire workflow ([workflow-syntax-... on docs.github](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore))
or conditions for some jobs ([using-conditions-... on docs.github](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution)) (see also examples for https://github.com/dorny/paths-filter)

**some (include) paths for ...**
compile and tests: src/, data/ (should we exclude some content?), test/, cmake/ (?), regression_test.py, (probably some more)
documentation: (only 1 min) doc/, src/, data/scripting/, data/tribes/, data/world/, data/txts/
clang-tidy: (2min 48s) src/
codecheck: (3min 17s) src/, CMakeLists.txt, */CMakeLists.txt
lua-style: (only 22s) data/campains/, data/i18n/, data/maps/, data/scripting/, data/templates/, data/tribes/, data/txts/, data/world/, test/ (or **.lua)

**notes**
To just skip **steps** or **jobs** if some files have changed, this github action could help: https://github.com/dorny/paths-filter
It also helps to only analyse changed files only, see https://github.com/dorny/paths-filter/tree/master#custom-processing-of-changed-files

**Reply by SimonHeimberg** [reply]
I can not push a change which modifies `.git/workflows/*` because of missing permission. So tried to attach a patch here. But I fail, the file format is not accepted, no matter if I create the file with `git show > xxx.patch` or `git format-patch`. (And PATCH is listed as supported file type in the error message telling this file type is not supported. :confused:)

Well, then I include the patch here:

~~~patch
From dee85abfae98c6195b93440fbe72095e1cc69065 Mon Sep 17 00:00:00 2001
From: Simon Heimberg <simon.heimberg@heimberg-ea.ch>
Date: Mon, 17 Jul 2023 18:08:35 +0200
Subject: [PATCH 1/2] only run tests and clang_tidy if a relevant file has
 changed

fixes #6020
enhances #5880
---
 .github/workflows/build.yaml | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/.github/workflows/build.yaml b/.github/workflows/build.yaml
index 290748fcbb..4ec7f0be87 100644
--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -81,6 +81,9 @@ jobs:
 
   clang_tidy:
     runs-on: ubuntu-latest
+    # Set job outputs to values from changed_files step for other jobs
+    outputs:
+      run_test_suite: ${{ steps.changed_files.outputs.test_suite }}
     env:
       CC:  /usr/bin/clang
       CXX: /usr/bin/clang++
@@ -89,9 +92,30 @@ jobs:
       uses: actions/checkout@v3
       with:
         fetch-depth: 1
+    - name: check for changed files
+      uses: dorny/paths-filter@v2
+      id: changed_files
+      with:
+        # groups of files to check for changed files
+        filters: |
+          clang_tidy:
+            - 'src/**'
+            - 'utils/run-clang-tidy.py'
+          compile: &compile
+            - 'CMakeLists.txt'
+            - 'cmake/**'
+            - 'src/**'
+          test_suite:
+            - *compile
+            - 'data/**'
+            - 'regression_test.py'
+            - 'test/**'
     - name: Installing dependencies
+      if: &run_clang_tidy steps.filter_files.outputs.clang_tidy == 'true'
       run: sh ./.github/scripts/install_deps.sh
     - name: CMake
+      # *run_clang_tidy is reference to above &run_clang_tidy
+      if: *run_clang_tidy
       run: |
         # Prepare build dir
         echo "Will use $(nproc) cores."
@@ -101,6 +125,7 @@ jobs:
         popd
     - name: Restore cache
       uses: actions/cache/restore@v3
+      if: *run_clang_tidy
       with:
         path: build/clang-tidy-cache
         key: clang-tidy-cache-${{ github.sha }}
@@ -109,6 +134,7 @@ jobs:
     - name: Check
       # To run a single check, call:
       # python3 ../utils/run-clang-tidy.py -checks=-*,my-check-prefix* | tee ../clang-tidy.log
+      if: *run_clang_tidy
       run: |
         # run-clang-tidy.py
         pushd build
@@ -116,11 +142,13 @@ jobs:
         popd
     - name: Save cache
       # We want to save the cache even if the report fails because of failed files
+      if: *run_clang_tidy
       uses: actions/cache/save@v3
       with:
         path: build/clang-tidy-cache
         key: clang-tidy-cache-${{ github.sha }}
     - name: Report
+      if: *run_clang_tidy
       run: utils/check_clang_tidy_results.py clang-tidy.log
 
   lua_style:
@@ -140,6 +168,7 @@ jobs:
 
   testsuite:
     needs: [clang_tidy, documentation, codecheck, misc_sanity_checks, lua_style]
+    if: needs.clang_tidy.outputs.run_test_suite == 'true'
     strategy:
       fail-fast: false
       matrix:
-- 
2.39.2
~~~

**Reply by Noordfrees** [reply]
I pushed the patch here: #6028

I also sent you an invitation to the dev team, then you should be able to push workflow changes

**Reply by SimonHeimberg** [reply]
This is in work and could be [done soon](https://github.com/widelands/widelands/pull/6257). Some tasks are in a separate thread/comment/something below.

**Comment by SimonHeimberg** [comment]
Some tasks from discussion in a pr (see links below):
- [x] Separate of each job in a separate workflow file (named build_XXX.yaml and called by build.yaml)
     https://github.com/widelands/widelands/pull/6257#discussion_r1414707020
- [x] For tests, divide data further. E.g. campaigns and most maps are not used in the tests. ...
      https://github.com/widelands/widelands/pull/6257#discussion_r1412805031
- [ ] Maybe we'll have to rethink whether we really want to create installers for all PR workflow runs, because that means almost half of the compilation jobs are always run.
      https://github.com/widelands/widelands/pull/6257#discussion_r1414721403

**Reply by tothxa** [reply]
- [ ] select 1 or 2 main setups from the testsuite matrix, and cache their executables, only run testsuite on these when only `data/` is changed
 - [x] ~only run tests from `test/maps/all_tribes.wmf` when only `data` is changed? or even only `test_<tribe>`, when only `data/tribes/*/<tribe>/*` is changed?~ *not needed any more after speeding the tests up*
